File: 0082.Remove-Duplicates-from-Sorted-List-II/step1.cpp
/**
 * Definition for singly-linked list.
 */
struct ListNode {
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};

class Solution {
  public:
    ListNode *deleteDuplicates(ListNode *head) {
        ListNode dummy = ListNode(0, head);
        ListNode *node = &dummy;
        while (node->next) {
            if (node->next->next && node->next->val == node->next->next->val) {
                int duplicated = node->next->val;
                while (node->next && node->next->val == duplicated) {
                    node->next = node->next->next;
                }
            }
            else {
                node = node->next;
            }
        }
        return dummy.next;
    }
};

File: 0021.Merge-Two-Sorted-Lists/step1.py
from typing import Optional


# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class Solution:
    def mergeTwoLists(
        self, list1: Optional[ListNode], list2: Optional[ListNode]
    ) -> Optional[ListNode]:
        node1 = list1
        node2 = list2
        dummy = ListNode()
        merged = dummy

        while node1 is not None or node2 is not None:
            if node1 is None:
                node1, node2 = node2, node1
            if node2 is None or node1.val < node2.val:
                merged.next = ListNode(val=node1.val)
                node1 = node1.next
            else:
                merged.next = ListNode(val=node2.val)
                node2 = node2.next
            merged = merged.next

        return dummy.next
